femwell/mode_solver.py

```diff
@@ -24,6 +29,105 @@
 from skfem.utils import solver_eigen_scipy
 
 
+@dataclass(frozen=True)
+class Modes:
+    modes: List
+
+    def __getitem__(self, idx):
+        return self.modes[idx]
+
+    def __len__(self):
+        return len(self.modes)
+
+    def __repr__(self) -> str:
+        modes = "\n\t" + "\n\t".join(repr(mode) for mode in self.modes) + "\n"
+        return f"{self.__class__.__name__}(modes=({modes}))"
+
+    def sorted(self, key):
+        return Modes(modes=sorted(self.modes, key=key))
+
+
+@dataclass(frozen=True)
+class Mode:
+    frequency: float
+    k: float
+    basis_epsilon_r: Basis
+    epsilon_r: NDArray
+    basis: Basis
+    E: NDArray
+    H: NDArray
+
+    @property
+    def omega(self):
+        return 2 * np.pi * self.frequency
+
+    @property
+    def k0(self):
+        return self.omega / speed_of_light
+
+    @property
+    def wavelength(self):
+        return speed_of_light / self.frequency
```

Comment:
hmm, I'd like a lot to do everything in SI-units, but it seems to lower the resolution. I'll investigate that further.
at the moment, we would use also another unit for frequency if we use non-si-units for wavelength

um -> GHz for example... what would here be the best way? I would want to keep `Mode` the same for electronics and photonics

Reply:
The simplest solution is to go dimensionless like is done in MEEP/MPB, see https://meep.readthedocs.io/en/latest/Introduction/#units-in-meep , or see table here https://github.com/abulnaga1/MPB_Simulations

Reply:
Sounds like a solution :)

My main worry would be, that there would always be a bit of confusion, how to put in numbers and also when it comes to calculation of derived values, e.g. the characteristic impedance.

What do you think? I somehow would like if it keeps close to "reality", but dimensionless is somehow elegant 🤔

Maybe there's also a way to have both?

Reply:
We could keep the mode solver dimensionless, and put all the unit manipulations and conversions outside.

For instance:

import femwell.units as units

# Defaults could be
mode = Mode(frequency=units.frequency(value=193, units="THz", length_units="um"))

where 

units.frequency(value=193, units="THz", length_units="um") = 0.67

but this could also do conversions like

units.frequency(value=193E3, units="GHz", length_units="um") = 0.67

Basically the required conversions happen in that utility module

The same could be done for outputs such as characteristic impedance. 

We could also go through Pint in that module.

Reply:
Sounds great!

The only problem I'd see here: what is the precision of calculations of derived values?

We would probably also need to keep those calculations dimensionless, too?

I think I'll first have a look if we can somehow get the mode solver not caring about dimensions, if that doesn't work, i think this would be a great solution?

Reply:
Okay, it seems now that we have it scale-invariant - i made sure, that the order of magnitude of E_x/y relative to E_z is scale invariant.
If this works generally, we can just stay with units :)